File: app/src/main/java/com/oreo/ui/custom/PercentageBar.java
package com.oreo.ui.custom;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.noisefit.luna.R;

public class PercentageBar extends View {

    private Paint paint;
    private int lineColor;
    private int percentageColor;
    private float interval;
    private int max;
    private int percentage;
    private float lineWidth = dp2px(1);
    private float cornerLine = dp2px(10);
    private float cornerPercentage = dp2px(2);

    private float intervalReal;
    private float percentageIndex;

    private int count;
    private RectF rectF;


    public PercentageBar(Context context) {
        super(context);
        initPaint();
    }

    public PercentageBar(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public PercentageBar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    public PercentageBar(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(attrs);
    }


    private void init(AttributeSet attrs) {
        TypedArray ta = getContext().obtainStyledAttributes(attrs, R.styleable.PercentageBar);
        lineColor = ta.getColor(R.styleable.PercentageBar_lineColor, Color.GRAY);
        percentageColor = ta.getColor(R.styleable.PercentageBar_percentageColor, Color.RED);
        interval = ta.getDimension(R.styleable.PercentageBar_interval, dp2px(2));
        max = ta.getInt(R.styleable.PercentageBar_max, 100);
        percentage = ta.getInt(R.styleable.PercentageBar_percentage, 50);
        lineWidth = ta.getDimension(R.styleable.PercentageBar_lineWidth, dp2px(2));
        ta.recycle();
        initPaint();
    }


    private void initPaint() {

        paint = new Paint();
        paint.setColor(lineColor);
        paint.setStyle(Paint.Style.FILL);
        paint.setStrokeWidth(lineWidth);
        paint.setAntiAlias(true);

        rectF = new RectF();
    }


    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);

        count = (int) Math.ceil((getWidth() / interval) / 2);

        intervalReal = getWidth() * 1f / count;

        percentageIndex = (int) Math.ceil(count * (percentage * 1f / max));


    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);
        drawContent(canvas);
    }

    private void drawContent(Canvas canvas) {
        Log.d("5555", "drawContent" + count + "  " + intervalReal);
        paint.setColor(lineColor);
        for (int i = 0; i < count; i++) {
            if (i > percentageIndex) {
                rectF.left = i * intervalReal;
                rectF.top = 0;
                rectF.right = rectF.left + lineWidth;
                rectF.bottom = getHeight();
                canvas.drawRoundRect(rectF, cornerLine, cornerLine, paint);
            }
        }

        paint.setColor(percentageColor);
        rectF.left = 0;
        rectF.top = 0;
        rectF.right = percentageIndex * intervalReal + lineWidth;
        rectF.bottom = getHeight();
        canvas.drawRoundRect(rectF, cornerPercentage, cornerPercentage, paint);
    }


    private int dp2px(float dpValue) {
        float scale = getContext().getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    public int getLineColor() {
        return lineColor;
    }

    public void setLineColor(int lineColor) {
        this.lineColor = lineColor;
    }

    public int getPercentageColor() {
        return percentageColor;
    }

    public void setPercentageColor(int percentageColor) {
        this.percentageColor = percentageColor;
    }

    public float getInterval() {
        return interval;
    }

    public void setInterval(float interval) {
        this.interval = interval;
    }

    public int getMax() {
        return max;
    }

    public void setMax(int max) {
        this.max = max;
    }

    public int getPercentage() {
        return percentage;
    }

    public void setPercentage(int percentage) {
        this.percentage = percentage;
    }

    public float getLineWidth() {
        return lineWidth;
    }

    public void setLineWidth(float lineWidth) {
        this.lineWidth = lineWidth;
    }

    public float getCornerLine() {
        return cornerLine;
    }

    public void setCornerLine(float cornerLine) {
        this.cornerLine = cornerLine;
    }

    public float getCornerPercentage() {
        return cornerPercentage;
    }

    public void setCornerPercentage(float cornerPercentage) {
        this.cornerPercentage = cornerPercentage;
    }
}

File: app/src/main/java/com/oreo/ui/custom/PercentageBar.kt
package com.oreo.ui.custom


import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import com.noisefit.luna.R
import com.noisefit_commans.utils.LOGS

class PercentageBar : View {
    private val intervalCount = 10
    private var lineColor = 0
    private var percentageColor = 0
    private var unitWidthLength = 0.0f
    private var rectF: RectF? = null
    private var cornerLine = dp2px(10f).toFloat()
    private var cornerProgressLine = dp2px(2f).toFloat()
    private var width = 0
    private val linePaint = Paint()
    private val percentagePaint = Paint()
    private var dataList = ArrayList<Int>()
    private var dataSize = 0
    private var isHighlighted = false

    constructor(context: Context?) : super(context) {
        initPaint()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(attrs)
    }

    constructor(
        context: Context?,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(attrs)
    }

    fun setPercentageColor1(percentageColor: Int) {
        this.percentageColor = percentageColor
    }

    private fun initPaint() {
        linePaint.apply {
            color = lineColor
            style = Paint.Style.FILL
            isAntiAlias = true
        }

        percentagePaint.apply {
            color = percentageColor
            style = Paint.Style.FILL
            isAntiAlias = true
        }

        rectF = RectF()
    }

    private fun init(attrs: AttributeSet?) {
        val ta = context.obtainStyledAttributes(attrs, R.styleable.PercentageBar)
        lineColor = ta.getColor(R.styleable.PercentageBar_lineColor, Color.GRAY)
        percentageColor = ta.getColor(R.styleable.PercentageBar_percentageColor, Color.RED)
//        interval = ta.getDimension(R.styleable.PercentageBar_interval, dp2px(2f).toFloat())
//        max = ta.getInt(R.styleable.PercentageBar_max, 100)
//        percentage = ta.getInt(R.styleable.PercentageBar_percentage, 50)
//        lineWidth = ta.getDimension(R.styleable.PercentageBar_lineWidth, dp2px(2f).toFloat())
        ta.recycle()
        initPaint()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        width = w
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        initPaint()
        drawBgLines(canvas)
        drawProgress(canvas)

    }

    private fun drawBgLines(canvas: Canvas?) {
        unitWidthLength = (width.toFloat() / (dataSize - 1))

        val dummyUnitLength = width / 100
        for (i in 0..200 step 2) {
            rectF!!.left = (i * dummyUnitLength).toFloat()
            rectF!!.top = 0f
            rectF!!.right = rectF!!.left + (dummyUnitLength)
            rectF!!.bottom = height.toFloat()
            canvas?.drawRoundRect(rectF!!, cornerLine, cornerLine, linePaint)
        }
    }


    fun updateData(
        dataList: List<Int>,
        dataSize: Int,
        color: Int,
        isHighlighted: Boolean
    ) {
        this.dataList.clear()
        this.dataList.addAll(dataList)
        this.dataSize = dataSize

        this.isHighlighted = isHighlighted
        this.percentageColor = color
        invalidate()
    }


    private fun drawProgress(canvas: Canvas?) {
        val data = getData()

        rectF!!.top = 0f
        rectF!!.bottom = height.toFloat()

        var left = 0f
        var hasLeft = false
        val halfUnitWidth = unitWidthLength / 2
        for (i in 0 until data.size) {
            val next = data.getOrNull(i + 1) ?: 0
            val current = data.getOrNull(i) ?: 0
            if (current == 1) {
                if (!hasLeft) {
                    left = if (i == 0) {
                        0f
                    } else {
                        (i * unitWidthLength) - unitWidthLength
                    }
                    hasLeft = true
                }
                if (i == dataSize - 1 || next == 0) {
                    rectF!!.left = left
                    rectF!!.right = if (i == 0) {
                        halfUnitWidth
                    } else {
                        (i * unitWidthLength)
                    }
                    canvas?.drawRoundRect(
                        rectF!!,
                        cornerProgressLine,
                        cornerProgressLine,
                        percentagePaint
                    )
                    hasLeft = false
                }
            }
        }


    }

    private fun dp2px(dpValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    private fun getData(): ArrayList<Int> {
        val data = ArrayList<Int>()
        if (isHighlighted) {
            for (i in 0 until dataSize) {
                if (dataList.contains(i)) {
                    data.add(1)
                } else {
                    data.add(0)
                }
            }
        } else {
            for (i in 0 until dataSize) {
                if (i < dataList.size) {
                    data.add(1)
                } else {
                    data.add(0)
                }
            }
        }

        return data
    }
}
